snapshots/windows/cimfs.go

```diff
@@ -0,0 +1,276 @@
+//go:build windows
+// +build windows
+
+/*
+   Copyright The containerd Authors.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package windows
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"path/filepath"
+
+	"github.com/Microsoft/hcsshim/pkg/cimfs"
+	cimlayer "github.com/Microsoft/hcsshim/pkg/ociwclayer/cim"
+	"github.com/containerd/containerd/errdefs"
+	"github.com/containerd/containerd/log"
+	"github.com/containerd/containerd/mount"
+	"github.com/containerd/containerd/plugin"
+	"github.com/containerd/containerd/snapshots"
+	"github.com/containerd/containerd/snapshots/storage"
+	"github.com/containerd/continuity/fs"
+)
+
+// Composite image FileSystem (CimFs) is a new read-only filesystem (similar to overlayFS on Linux) created
+// specifically for storing container image layers on windows.  cimfsSnapshotter is a snapshotter that uses
+// CimFs to create read-only parent layer snapshots. Each snapshot is represented by a `<snapshot-id>.cim`
+// file and some other files which hold contents of that snapshot.  Once a cim file for a layer is created it
+// can only be used as a read-only layer by mounting it to a volume. Hence, CimFs will not be used when we are
+// creating writable layers for container scratch and such. (However, scratch layer of a container can be
+// exported to a cim layer and then be used as a parent layer for another container).  Since CimFs can not be
+// used for scratch layers we still use some functionality of the existing windows snapshotter to create
+// writable scratch space snapshots.
```

Comment:
It seems awkard to include the "legacy" snapshotter inside of the new CimFS snapshotter. If we just need this for the scratch VHD handling functionality, wouldn't it be simpler to break that out into its own set of functions that both snapshotters use?

If there really is a lot in common between both snapshotters, it may be better to implement a "common" snapshotter that both wrap.

Reply:
Fixed this to have a common base snapshotter.